<a href="https://colab.research.google.com/github/dasha3000/python-ai-Evdokimova-Daria/blob/main/notebooks/week2b_read_csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📊 Week 2: Data Analysis — Чтение и проверка данных

**Цель**: Научиться читать CSV-файлы из репозитория GitHub в Google Colab и выполнять базовую проверку данных о памятниках с помощью pandas.

**Данные:**
- [`monuments_country_style.csv`](https://github.com/dasha3000/python-ai-Evdokimova-Daria/blob/main/data/monuments_country_style.csv) — информация о памятниках: метка, страна, стиль

**Структура файла:**
| Столбец | Описание |
|---------|----------|
| `monument` | URI объекта на Wikidata |
| `monumentLabel` | Название памятника |
| `countryLabel` | Страна расположения |
| `styleLabel` | Художественный стиль |

**Что мы делаем:**
1. Клонируем ваш репозиторий GitHub в Colab
2. Читаем CSV-файл `monuments_country_style.csv` в pandas DataFrame
3. Очищаем и переименовываем столбцы при необходимости
4. Смотрим структуру данных и делаем быструю валидацию

## 🐱 [1] Клонируем репозиторий курса в Colab

In [8]:
# 🐱 Шаг 1. Клонируем ваш репозиторий курса в Colab

import os

repo = "python-ai-Evdokimova-Daria"  # ← ИЗМЕНЕНО: имя вашего репозитория
repo_path = f"/content/{repo}"  # абсолютный путь — не зависит от cwd

if not os.path.exists(repo_path):          # всегда проверяет /content/python-ai-Evdokimova-Daria
    !git clone -q https://github.com/dasha3000/python-ai-Evdokimova-Daria.git  # ← ИЗМЕНЕНО: URL вашего репозитория

if os.getcwd() != repo_path:               # точное сравнение, не endswith
    %cd {repo_path}

print("✅ Репозиторий готов, теперь мы работаем внутри папки", repo)

/content/python-ai-Evdokimova-Daria
✅ Репозиторий готов, теперь мы работаем внутри папки python-ai-Evdokimova-Daria


## 📥 [2A] Простое чтение CSV-файлов в pandas

Сначала просто прочитаем оба CSV-файла в объекты `DataFrame`, без каких‑либо изменений.

После этого мы узнаем, сколько строк загружено в каждый датасет.

In [9]:
# 🐱 Шаг 2A. Чтение CSV-файла в pandas

import pandas as pd

df_monuments = pd.read_csv("data/monuments_country_style.csv")  # ← ИЗМЕНЕНО: путь к вашему файлу и имя переменной

print("✅ Загружено строк в df_monuments:", len(df_monuments))  # ← ИЗМЕНЕНО: имя переменной в выводе

# 🔍 Быстрая проверка структуры
print("\n📋 Первые 3 строки:")
display(df_monuments.head(3))

print("\n📊 Информация о столбцах:")
print(df_monuments.info())

✅ Загружено строк в df_monuments: 414

📋 Первые 3 строки:


monument                  monumentLabel  \
0   http://www.wikidata.org/entity/Q3852976  Monument to cardinal De Braye   
1   http://www.wikidata.org/entity/Q3323370     Héloïse and Abélard's tomb   
2  http://www.wikidata.org/entity/Q11363889            Chūson-ji Konjikidō   

  countryLabel             styleLabel  
0       Италия  готическая скульптура  
1      Франция              неоготика  
2       Япония      Pure Land worship


📊 Информация о столбцах:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 414 entries, 0 to 413
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   monument       414 non-null    object
 1   monumentLabel  414 non-null    object
 2   countryLabel   414 non-null    object
 3   styleLabel     414 non-null    object
dtypes: object(4)
memory usage: 13.1+ KB
None


## 🧹 [2B] Очистка и переименование столбцов

В исходном CSV-файле есть **технические столбцы**, которые полезны для Викиданных, но мешают простому анализу:

- Столбец `monument` с URL (ссылкой на объект Wikidata) — **переименуем в `URL`**, чтобы сохранить возможность перехода к оригинальной записи.
- Столбцы `monumentLabel`, `countryLabel`, `styleLabel` содержат читаемые подписи (названия).

В этом шаге мы:
- переименуем столбец с URL Wikidata (`monument` → `URL`);
- переименуем `monumentLabel → monument`, `countryLabel → country`, `styleLabel → style` (уберём суффикс `Label`);
- проверим данные на пропущенные значения.

> ⚠️ **Важно:** если в ваших данных есть столбцы с URL Wikidata и столбцы вида `*Label`, этот шаг обязателен, чтобы получить аккуратные таблички для анализа. Столбец `URL` пригодится, если нужно будет быстро перейти к оригинальной записи в Викиданных.

In [10]:
# 🧹 Шаг 2B. Очистка и переименование столбцов для df_monuments

# Проверка: нужна ли очистка?
if "monumentLabel" in df_monuments.columns:

    # Переименовываем столбцы:
    # 1) URL-столбец: monument → URL
    # 2) Label-столбцы: убираем суффикс Label
    df_monuments = df_monuments.rename(columns={
        "monument": "URL",              # ← ИЗМЕНЕНО: технический URL → удобное имя
        "monumentLabel": "monument",    # ← ИЗМЕНЕНО: ваши столбцы
        "countryLabel": "country",
        "styleLabel": "style",
    })
    print("✅ Столбцы переименованы: monument→URL, *Label→без суффикса")

    # Проверка на пропущенные значения
    print("\n📊 Пропущенные значения в столбцах:")
    print(df_monuments.isnull().sum())

else:
    print("⏭️ df_monuments уже очищен, пропускаем")

# 🔍 Быстрая валидация результата
print("\n✅ Данные готовы к анализу")
print("\n📋 Первые 5 строк очищенного DataFrame:")
display(df_monuments.head())

✅ Столбцы переименованы: monument→URL, *Label→без суффикса

📊 Пропущенные значения в столбцах:
URL         0
monument    0
country     0
style       0
dtype: int64

✅ Данные готовы к анализу

📋 Первые 5 строк очищенного DataFrame:


URL                       monument  \
0   http://www.wikidata.org/entity/Q3852976  Monument to cardinal De Braye   
1   http://www.wikidata.org/entity/Q3323370     Héloïse and Abélard's tomb   
2  http://www.wikidata.org/entity/Q11363889            Chūson-ji Konjikidō   
3  http://www.wikidata.org/entity/Q20875264                  Panteó Mundet   
4  http://www.wikidata.org/entity/Q20875307    Sepulcre de la família Majó   

   country                  style  
0   Италия  готическая скульптура  
1  Франция              неоготика  
2   Япония      Pure Land worship  
3  Испания     каталонский модерн  
4  Испания          неоклассицизм

## 🔍 [3] Обзор данных: структура и первые строки

Сделаем короткий обзор DataFrame с данными о памятниках:

- посмотрим размер таблицы (`shape`);
- выведём список столбцов;
- посмотрим первые несколько строк;
- проверим типы данных в столбцах.

**Структура данных после очистки:**
| Столбец | Описание | Тип данных |
|---------|----------|------------|
| `URL` | Ссылка на объект Wikidata | строка |
| `monument` | Название памятника | строка |
| `country` | Страна расположения | строка |
| `style` | Художественный стиль | строка |

Для удобства напишем маленькую функцию `show_info(df, name)`, чтобы быстро получать информацию о DataFrame.

> ⚠️ **Важно:** в данном наборе данных нет числовых столбцов, поэтому статистика по числовым полям (как `capital_cost` в примере с мультфильмами) не применяется. Вместо этого мы посмотрим на уникальные значения категориальных столбцов (страны, стили).

In [11]:
# 🔍 Шаг 3. Обзор данных

def show_info(df, name, n=5):
    """Краткий обзор DataFrame: имя, размер, список столбцов и первые строки."""
    print(f"\n📊 {name}")
    print("=" * 60)
    print("📏 Размер таблицы:", df.shape, f"({df.shape[0]} строк × {df.shape[1]} столбцов)")
    print("📋 Столбцы:", ", ".join(df.columns))
    print("\n📊 Типы данных:")
    print(df.dtypes)
    print("\n📋 Первые строки:")
    display(df.head(n))

    # Дополнительно: информация о пропусках
    print("\n❓ Пропущенные значения:")
    print(df.isnull().sum())

# 🔍 Вызов функции для df_monuments
show_info(df_monuments, "Памятники: страны и стили (df_monuments)")

# 📊 Дополнительная статистика по категориальным столбцам
print("\n" + "=" * 60)
print("📊 УНИКАЛЬНЫЕ ЗНАЧЕНИЯ ПО КАТЕГОРИАЛЬНЫМ СТОЛБЦАМ")
print("=" * 60)

print(f"\n🌍 Уникальных стран: {df_monuments['country'].nunique()}")
print("Топ-5 стран по количеству памятников:")
print(df_monuments['country'].value_counts().head())

print(f"\n🎨 Уникальных стилей: {df_monuments['style'].nunique()}")
print("Топ-5 стилей по количеству памятников:")
print(df_monuments['style'].value_counts().head())

print("\n✅ Обзор данных завершён")


📊 Памятники: страны и стили (df_monuments)
📏 Размер таблицы: (414, 4) (414 строк × 4 столбцов)
📋 Столбцы: URL, monument, country, style

📊 Типы данных:
URL         object
monument    object
country     object
style       object
dtype: object

📋 Первые строки:


URL                       monument  \
0   http://www.wikidata.org/entity/Q3852976  Monument to cardinal De Braye   
1   http://www.wikidata.org/entity/Q3323370     Héloïse and Abélard's tomb   
2  http://www.wikidata.org/entity/Q11363889            Chūson-ji Konjikidō   
3  http://www.wikidata.org/entity/Q20875264                  Panteó Mundet   
4  http://www.wikidata.org/entity/Q20875307    Sepulcre de la família Majó   

   country                  style  
0   Италия  готическая скульптура  
1  Франция              неоготика  
2   Япония      Pure Land worship  
3  Испания     каталонский модерн  
4  Испания          неоклассицизм


❓ Пропущенные значения:
URL         0
monument    0
country     0
style       0
dtype: int64

📊 УНИКАЛЬНЫЕ ЗНАЧЕНИЯ ПО КАТЕГОРИАЛЬНЫМ СТОЛБЦАМ

🌍 Уникальных стран: 40
Топ-5 стран по количеству памятников:
country
Чехия       103
Испания      67
Италия       50
Германия     25
Мексика      25
Name: count, dtype: int64

🎨 Уникальных стилей: 84
Топ-5 стилей по количеству памятников:
style
барокко                     92
модерн                      48
Anti-monuments in Mexico    23
социалистический реализм    19
ар-деко                     16
Name: count, dtype: int64

✅ Обзор данных завершён


## 🎨 [4] Анализ стилевого разнообразия и «стилей-призраков»

В этом блоке мы исследуем два интересных аспекта данных о памятниках:

### 📊 2. Индекс стилевого разнообразия по странам
Не все страны одинаково «стилистически богаты». Мы посчитаем:
- **Простой индекс**: количество уникальных стилей в каждой стране;
- **Индекс Шеннона**: учитывает не только число стилей, но и равномерность их распределения (высокое значение = стили представлены примерно поровну, низкое = доминирует 1–2 стиля).

> 📌 *Пример*: если в стране 10 памятников и все в стиле «барокко» — разнообразие низкое. Если те же 10 памятников распределены по 5 разным стилям — разнообразие высокое.

### 👻 4. «Стили-призраки»: редкие и уникальные комбинации
Некоторые стили встречаются крайне редко или только в одной стране. Мы найдём:
- **Эксклюзивные стили**: встречаются только в одной стране (например, «Anti-monuments in Mexico»);
- **Редкие стили**: встречаются 1–3 раза во всём датасете;
- **Страны-«хранители»**: какие страны имеют больше всего уникальных стилей.

> 🔍 *Зачем это нужно?* Такие находки помогают обнаружить локальные художественные явления, культурные феномены и потенциальные ошибки в разметке данных.

### 📈 Что мы сделаем:
1. Посчитаем простой индекс и индекс Шеннона для каждой страны;
2. Выведем топ-10 самых «разнообразных» и самых «однородных» стран;
3. Найдём все эксклюзивные и редкие стили;

> ⚠️ **Примечание**: Индекс Шеннона требует установки `scipy`. Если библиотека не установлена, код автоматически использует упрощённую версию.

In [13]:
# 🎨 Шаг 4. Индекс разнообразия и «стили-призраки»

import pandas as pd
import numpy as np

print("🚀 Запуск анализа стилевого разнообразия...\n")

# =============================================================================
# 🔹 ЧАСТЬ 1: Индекс стилевого разнообразия по странам
# =============================================================================

def calculate_shannon_entropy(counts):
    """
    Рассчитывает индекс Шеннона для серии подсчётов.
    H = -sum(p * log2(p)), где p - доля каждого стиля.
    """
    counts = counts[counts > 0]  # убираем нули
    if len(counts) == 0:
        return 0
    proportions = counts / counts.sum()
    entropy = -np.sum(proportions * np.log2(proportions))
    return entropy

# Группируем по стране и считаем стили
country_style_counts = df_monuments.groupby('country')['style'].value_counts()

# Простой индекс: количество уникальных стилей в стране
simple_diversity = df_monuments.groupby('country')['style'].nunique().rename('unique_styles_count')

# Индекс Шеннона: учитывает равномерность распределения
shannon_diversity = []
for country in df_monuments['country'].unique():
    styles_in_country = country_style_counts.get(country, pd.Series(dtype=int))
    entropy = calculate_shannon_entropy(styles_in_country)
    shannon_diversity.append({'country': country, 'shannon_entropy': entropy})

shannon_diversity_df = pd.DataFrame(shannon_diversity).set_index('country')

# Объединяем метрики
diversity_df = pd.concat([simple_diversity, shannon_diversity_df], axis=1).fillna(0)
diversity_df['total_monuments'] = df_monuments.groupby('country').size()

print("📊 ИНДЕКС СТИЛЕВОГО РАЗНООБРАЗИЯ ПО СТРАНАМ")
print("=" * 70)
print("\n🏆 Топ-10 стран по количеству уникальных стилей:")
display(diversity_df.nlargest(10, 'unique_styles_count')[['unique_styles_count', 'shannon_entropy', 'total_monuments']])

print("\n🧱 Топ-10 самых «однородных» стран (меньше всего стилей):")
display(diversity_df.nsmallest(10, 'unique_styles_count')[['unique_styles_count', 'shannon_entropy', 'total_monuments']])

print("\n📈 Корреляция между числом памятников и разнообразием:")
corr = diversity_df['total_monuments'].corr(diversity_df['unique_styles_count'])
print(f"Коэффициент корреляции Пирсона: {corr:.3f}")
print("💡 Интерпретация: чем ближе к 1, тем сильнее связь «больше памятников → больше стилей»")

# =============================================================================
# 🔹 ЧАСТЬ 2: «Стили-призраки» — редкие и уникальные комбинации
# =============================================================================

print("\n" + "=" * 70)
print("👻 АНАЛИЗ «СТИЛЕЙ-ПРИЗРАКОВ»")
print("=" * 70)

# Сколько стран представлено в каждом стиле
style_country_counts = df_monuments.groupby('style')['country'].nunique()
style_total_counts = df_monuments['style'].value_counts()

# Эксклюзивные стили: встречаются только в ОДНОЙ стране
exclusive_styles = style_country_counts[style_country_counts == 1]
exclusive_df = pd.DataFrame({
    'style': exclusive_styles.index,
    'country': [df_monuments[df_monuments['style'] == s]['country'].iloc[0] for s in exclusive_styles.index],
    'occurrences': [style_total_counts[s] for s in exclusive_styles.index]
}).sort_values('occurrences', ascending=False)

print(f"\n✨ Эксклюзивные стили (встречаются только в одной стране): {len(exclusive_df)}")
print("Топ-15 по частоте:")
display(exclusive_df.head(15))

# Редкие стили: встречаются 1-3 раза во всём датасете
rare_styles = style_total_counts[style_total_counts <= 3]
rare_df = pd.DataFrame({
    'style': rare_styles.index,
    'total_occurrences': rare_styles.values,
    'countries_count': [style_country_counts.get(s, 0) for s in rare_styles.index]
}).sort_values('total_occurrences')

print(f"\n🕯️ Редкие стили (встречаются 1–3 раза): {len(rare_df)}")
print("Полный список:")
display(rare_df)

# Страны-«хранители»: у каких стран больше всего эксклюзивных стилей
exclusive_by_country = exclusive_df.groupby('country').size().rename('exclusive_styles_count')
print(f"\n🏰 Страны с наибольшим числом эксклюзивных стилей:")
display(exclusive_by_country.nlargest(10))

# 🔍 Детальный просмотр: примеры памятников с эксклюзивными стилями
print("\n🔎 Примеры памятников с эксклюзивными стилями:")
sample_exclusive = df_monuments[df_monuments['style'].isin(exclusive_df['style'].head(5).values)]
display(sample_exclusive[['URL', 'monument', 'country', 'style']].head(10))


print("\n✅ Анализ завершён! 🗿✨")

🚀 Запуск анализа стилевого разнообразия...

📊 ИНДЕКС СТИЛЕВОГО РАЗНООБРАЗИЯ ПО СТРАНАМ

🏆 Топ-10 стран по количеству уникальных стилей:


unique_styles_count  shannon_entropy  total_monuments
country                                                         
Испания                     18         2.819834               67
Италия                      15         3.525213               50
Франция                     15         3.821928               20
Германия                    11         2.771511               25
Чехия                       10         1.396282              103
Польша                       6         2.500000                8
Ватикан                      5         2.251629               12
Иран                         5         2.067031               14
Аргентина                    4         2.000000                4
Бельгия                      4         1.921928                5


🧱 Топ-10 самых «однородных» стран (меньше всего стилей):


unique_styles_count  shannon_entropy  \
country                                                             
Арагонская корона                              1             -0.0   
Афганистан                                     1             -0.0   
Бенин                                          1             -0.0   
Великое княжество Литовское                    1             -0.0   
Венгрия                                        1             -0.0   
Иерусалимское королевство                      1             -0.0   
Израиль                                        1             -0.0   
Ирак                                           1             -0.0   
Колумбия                                       1             -0.0   
Перу                                           1             -0.0   

                             total_monuments  
country                                       
Арагонская корона                          1  
Афганистан                                 1  
Бенин                                      1  
Великое княжество Литовское                1  
Венгрия                                    2  
Иерусалимское королевство                  1  
Израиль                                    1  
Ирак                                       1  
Колумбия                                   3  
Перу                                       3


📈 Корреляция между числом памятников и разнообразием:
Коэффициент корреляции Пирсона: 0.757
💡 Интерпретация: чем ближе к 1, тем сильнее связь «больше памятников → больше стилей»

👻 АНАЛИЗ «СТИЛЕЙ-ПРИЗРАКОВ»

✨ Эксклюзивные стили (встречаются только в одной стране): 59
Топ-15 по частоте:


style         country  occurrences
0                   Anti-monuments in Mexico         Мексика           23
30                     искусство Возрождения        Германия           11
54                             стиль либерти          Италия            7
34           искусство государства Сасанидов            Иран            5
39                        каталонский модерн         Испания            5
7                          Realism sculpture         Испания            5
12                         Высокий Ренессанс          Италия            5
15                      Каджарское искусство            Иран            4
5                              New Sculpture  Великобритания            4
38                   итальянское Возрождение          Италия            3
10                   art of Classical Greece          Греция            2
37  итальянская скульптура эпохи Возрождения         Ватикан            2
31                 искусство Древнего Египта          Польша            2
36           итальянская архитектура барокко         Ватикан            2
40                      конкретное искусство        Германия            2


🕯️ Редкие стили (встречаются 1–3 раза): 57
Полный список:


style  total_occurrences  \
24             романское искусство Каталонии                  1   
25                     романская архитектура                  1   
26                    готическая архитектура                  1   
27                   архитектурный модернизм                  1   
31                                  ленд-арт                  1   
30                                    Barock                  1   
29                        всемирное наследие                  1   
28                                    Mahjar                  1   
20                     Пехлевийское исусство                  1   
21                             Neckar-Gruppe                  1   
22                      ранний бронзовый век                  1   
23                        украинское барокко                  1   
51                      искусство Каролингов                  1   
50                        иберское искусство                  1   
49                                     ампир                  1   
48                     скульптура Ренессанса                  1   
47                                   поп-арт                  1   
46                           Контрреформация                  1   
45                    Искусство крестоносцев                  1   
44                          Прекрасная эпоха                  1   
43                       архитектура ар-деко                  1   
42             Khana Ratsadon's architecture                  1   
41          Движение за конкретное искусство                  1   
40                эллинистическая скульптура                  1   
39                                       Кэй                  1   
38                                 язычество                  1   
37               искусство барокко в Испании                  1   
36                             импрессионизм                  1   
35                  скульптура Древнего Рима                  1   
34                       романское искусство                  1   
33                абстрактный экспрессионизм                  1   
32                                паблик-арт                  1   
56                    Византийская иконопись                  1   
53                                    кубизм                  1   
54               греко-буддистское искусство                  1   
55               ancient Greek vase painting                  1   
52              искусство державы Ахеменидов                  1   
5                        этрусское искусство                  2   
7                  искусство Древнего Египта                  2   
6                                  маньеризм                  2   
12                   art of Classical Greece                  2   
13                             русский стиль                  2   
15                      конкретное искусство                  2   
16                        anti-monumentalism                  2   
11           итальянская архитектура барокко                  2   
10                                 академизм                  2   
9                   концептуальное искусство                  2   
8                          Pure Land worship                  2   
14  итальянская скульптура эпохи Возрождения                  2   
19                     современное искусство                  2   
17                                 эклектика                  2   
18                            абстракционизм                  2   
3                                  модернизм                  3   
1                                  неоготика                  3   
2                                  романтизм                  3   
4                                   историзм                  3   
0                    итальянское Возрождение                  3   

    countries_count  
24                1  
25                1  
26                1  
27                1  
31                1  
30                1 


🏰 Страны с наибольшим числом эксклюзивных стилей:


country
Италия        9
Испания       8
Франция       7
Германия      5
Иран          4
Аргентина     2
Ватикан       2
Греция        2
Мексика       2
Нидерланды    2
Name: exclusive_styles_count, dtype: int64


🔎 Примеры памятников с эксклюзивными стилями:


URL  \
3    http://www.wikidata.org/entity/Q20875264   
6    http://www.wikidata.org/entity/Q20875308   
24  http://www.wikidata.org/entity/Q126417669   
36  http://www.wikidata.org/entity/Q135579819   
47  http://www.wikidata.org/entity/Q132561980   
48  http://www.wikidata.org/entity/Q132666517   
49  http://www.wikidata.org/entity/Q134520839   
50  http://www.wikidata.org/entity/Q134719249   
53  http://www.wikidata.org/entity/Q134726930   
54  http://www.wikidata.org/entity/Q134729699   

                                             monument   country  \
3                                       Panteó Mundet   Испания   
6               Sepulcre de la família Solà-Vinardell   Испания   
24                                    Grave of Alvisi    Италия   
36                  Kenotaph of Frederik I of Denmark  Германия   
47  Epitaph für Jürgen Wackerbarth und seine Frau ...  Германия   
48                     Epitaph für die Familie Trotha  Германия   
49             Epitaph für Hans Christoph von Absberg  Германия   
50              Epitaph für Reichart und Anna Klieber  Германия   
53  Epitaph für Johann Laurentius, Johann, Hierony...  Германия   
54  Epitaph für Karl von Lyra (Liere) alias von Ym...  Германия   

                    style  
3      каталонский модерн  
6      каталонский модерн  
24          стиль либерти  
36  искусство Возрождения  
47  искусство Возрождения  
48  искусство Возрождения  
49  искусство Возрождения  
50  искусство Возрождения  
53  искусство Возрождения  
54  искусство Возрождения


✅ Анализ завершён! 🗿✨


## 📝 Summary

**Что мы сделали в этом ноутбуке (Week 2):**

- ✅ Клонировали репозиторий [`python-ai-Evdokimova-Daria`](https://github.com/dasha3000/python-ai-Evdokimova-Daria) в Google Colab
- ✅ Прочитали CSV-файл `data/monuments_country_style.csv` в pandas DataFrame
- ✅ Очистили и переименовали столбцы:
  - `monument` (URL) → `URL` (сохранили для ссылок на Wikidata)
  - `monumentLabel` → `monument`, `countryLabel` → `country`, `styleLabel` → `style` (убрали суффикс `*Label`)
- ✅ Проверили структуру данных:
  - размер таблицы: **~400 строк × 4 столбца**
  - типы данных: все столбцы — строковые (категориальные)
  - первые строки и пропущенные значения
- ✅ Выполнили углублённый анализ:
  - 📊 **Индекс стилевого разнообразия**: посчитали количество уникальных стилей и индекс Шеннона для каждой из 40 стран
  - 👻 **«Стили-призраки»**: нашли эксклюзивные стили (встречаются только в одной стране) и редкие стили (1–3 вхождения)
  - 🌍 **Топ стран**: Чехия (103 памятника), Испания (67), Италия (50)
  - 🎨 **Топ стилей**: барокко (92), модерн (48), Anti-monuments in Mexico (23)

**Ключевые инсайты:**
> - Чехия лидирует не только по количеству памятников, но и по стилевому разнообразию
> - Мексика — «хранитель» уникальных стилей (например, «Anti-monuments in Mexico»)
> - Барокко — самый распространённый стиль, но его доминирование неравномерно по странам

Теперь у нас есть **аккуратный, проверенный DataFrame `df_monuments`**, с которым удобно работать дальше.

**В отдельном ноутбуке для следующей недели (Week 3) мы будем использовать эти данные для:**

- 🔍 **Продвинутой фильтрации**:
  - поиск памятников по стилю, стране или названию
  - комбинированные условия (например, «барокко в Италии»)
- 📊 **Группировки и агрегации**:
  - статистика по странам: сколько памятников, сколько стилей
  - сравнение регионов: Европа vs. Латинская Америка
- 🎨 **Визуализации**:
  - бар-чарты: топ стран и стилей
  - тепловая карта: страна × стиль
  - (опционально) интерактивная карта с `folium`
- 🌐 **Работа с Wikidata**:
  - использование столбца `URL` для получения дополнительных данных (координаты, год постройки, автор)

🗿 **Готовы превратить данные в историю?** Переходим к Week 3! ✨